tutorials/yes-no-game/full game/yesnogame.edn

```diff
@@ -0,0 +1,251 @@
+
+; creates an UI.Area with the left image
+(defshards left-image [image]
+  (UI.Area
+   :Position (float2 left-x-offset, 0.0)
+   :Anchor Anchor.Center
+   :Contents
+   (-> image (UI.Image :Scale (float2 0.5)))))
+
+; creates an UI.Area with the right image
+(defshards right-image [image]
+  (UI.Area
+   :Position (float2 right-x-offset, 0.0)
+   :Anchor Anchor.Center
+   :Contents
+   (-> image (UI.Image :Scale (float2 0.5)))))
+
+; creates the main game area
+(defshards main-game-ui []
+  (UI.TopPanel
+   (->
+    (UI.Horizontal
+     (->
+      "Score: " (UI.Label)
+      .total-score (ToString) (UI.Label)
+      (UI.Separator)
+      "Round: " (UI.Label)
+      .current-round (ToString) (UI.Label)
+      (UI.Separator)
+      "Time Left: " (UI.Label)
+      max-timer(time-remaining)(ToString)(UI.Label)))))
+
+  (UI.BottomPanel
+   (-> "Are they the same type of animal? Press the UP arrow if YES, and the DOWN arrow if NO." (UI.Label)))
+  
+  (UI.CentralPanel
+   (->
+    (UI.Horizontal
+     :Wrap true
+     :Contents
+     (->
+      (left-image (-> .all-images (Take .left-series-index) (Take .left-image-index)))
+      (right-image (-> .all-images (Take .right-series-index) (Take .right-image-index))))))))
+
+; creates the game over screen
+(defshards game-over-ui []
+  (UI.CentralPanel
+   (->
+    (UI.Area
+     :Position (float2 0.0, 0.0)
+     :Anchor Anchor.Center
+     :Contents
+     (->
+      "GAME OVER" (UI.Label)
+      (UI.Horizontal
+       :Contents
+       (->
+        "Final Score: " (UI.Label)
+        .total-score (ToString) (UI.Label)
+        "/"  (UI.Label)
+        total-rounds (ToString) (UI.Label)))
+
+      (UI.Button
+       :Label "Play Again!"
+       :Action (-> (reset-values))))))))
+
+(defshards check-answer [yes-input]
+  (When
+   :Predicate (-> .input-received (Is false))
+   :Action (-> 
+          (When
+           :Predicate (-> .same-images (Is yes-input))
+           :Action (-> (Math.Inc .total-score))) )) 
+  true > .input-received
+  (end-round))
+
+(defshards end-round []
+  .current-round (Math.Add 1) > .new-round-number
+  (If
+   :Predicate (-> .new-round-number (IsMore total-rounds))
+   :Then (->
+          true > .game-over)
+   :Else (->
+          .new-round-number > .current-round
+          true > .new-round)))
+
+(defshards increase-timer []
+  (When
+   :Predicate(-> .game-over(Is false))
+   :Action(->
+           "Increasing Timer"(Log)
+           .seconds-elapsed(Log)
+           (Math.Inc .seconds-elapsed)
+           (When
+            :Predicate (-> .seconds-elapsed (IsMore max-timer))
+            :Action (end-round)))))
+
+(defn time-remaining []
+  (Math.Subtract .seconds-elapsed))
+
+(defloop game-loop
+  (Setup (initialize))
+
+  (Once
+   :Action (-> (increase-timer))
+   :Every 1.0)
+  
+  (When
+   :Predicate  (-> (IsNot .game-over))
+   :Action (->
+          (When
+           :Predicate (-> .new-round)
+           :Action (-> (initialize-round)))))
```

Comment:
Prefer to put long shards on the next line for better readability.

Additionally, it can be written a bit differently considering `WhenNot` is the reverse logic of `When`.
```suggestion  
  (WhenNot
   :Predicate  (-> .game-over)
   :Action
   (When
    :Predicate (-> .new-round)
    :Action (initialize-round)))
```